### PR TITLE
Issue 14382 - converting old D1 operator overloading style to new D2 style

### PR DESCRIPTION
**Pull request by schuetzm**
Patches by ketmar: https://issues.dlang.org/show_bug.cgi?id=14382

### CODE REVIEW

**Comment by Safety0ff** [comment]
Previous discussions for reference:
https://github.com/D-Programming-Language/phobos/pull/2093
http://forum.dlang.org/thread/ifghzjafvfqrqkhlpinc@forum.dlang.org
https://issues.dlang.org/show_bug.cgi?id=12462

File: std/variant.d [hunk 1]
```diff
@@ -1025,50 +1027,50 @@ public:
...
-    VariantN opShr(T)(T rhs) { return opLogic!(T, ">>")(rhs); }
+    VariantN opBinary(string op : ">>", T)(T rhs) { return opLogic!(T, ">>")(rhs); }
     // ///ditto
     // VariantN opShr_r(T)(T lhs)
     // {
     //     return VariantN(lhs).opLogic!(VariantN, ">>")(this);
     // }
     ///ditto
-    VariantN opUShr(T)(T rhs) { return opLogic!(T, ">>>")(rhs); }
+    VariantN opBinary(string op : ">>>", T)(T rhs) { return opLogic!(T, ">>>")(rhs); }
```
Comment (schveiguy) [inline std/variant.d]: All these functions are nearly identical. The whole point of having the template was so you can easily implement multiple operators with one set of code. For example:

```D
VariantN opBinary(string op, T)(T rhs) if (op == "+" || op == "-" || op = "*" || op = "/")
{
    return opArithmetic!(T, op)(rhs);
}
```

In fact, I'd say get rid of `opArithmetic` and `opLogic`, and just move the implementations into an appropriate `opBinary`

File: std/variant.d [hunk 2]
```diff
@@ -1083,29 +1085,29 @@ public:
...
+    VariantN opOpAssign(string op : "^", T)(T rhs)  { return this = this ^ rhs; }
     ///ditto
-    VariantN opShlAssign(T)(T rhs)  { return this = this << rhs; }
+    VariantN opOpAssign(string op : "<<", T)(T rhs)  { return this = this << rhs; }
     ///ditto
-    VariantN opShrAssign(T)(T rhs)  { return this = this >> rhs; }
+    VariantN opOpAssign(string op : ">>", T)(T rhs)  { return this = this >> rhs; }
     ///ditto
-    VariantN opUShrAssign(T)(T rhs) { return this = this >>> rhs; }
+    VariantN opOpAssign(string op : ">>>", T)(T rhs) { return this = this >>> rhs; }
```
Comment (schveiguy) [inline std/variant.d]: Replace all these with this:

```D
VariantN opOpAssign(string op, T)(T rhs) if (/* op is one of these things */)
{
   mixin("return this = this " ~ op ~ " rhs;");
}
```

**Comment by yebblies** [comment]
This makes the code less readable.  The old operator overloading methods aren't deprecated.  The new methods are more powerful, and that power is wasted in 100% of these cases.

**Comment by schveiguy** [comment]
> This makes the code less readable.

What about with my suggested updates?

> The old operator overloading methods aren't deprecated.

The old operator overloading methods are not documented (I will correct this), but should not be used in Phobos IMO. If there was ever a type that could use template + mixin opBinary, it would be Variant.

**Comment by DmitryOlshansky** [comment]
Seems useful if @schveiguy suggestions are followed through. @schuetzm any plans to update this?

**Comment by schuetzm** [comment]
This is ketmar's patch, I'm going to ask him. Maybe I'll also find some time to do it myself.

**Comment by JakobOvrum** [comment]
@schuetzm, any news on this?

**Comment by quickfur** [comment]
ping
Are we going ahead with this, or should we close it?

**Comment by schveiguy** [comment]
@quickfur I don't think it's worth merging this unless we take advantage of the templated nature of the new operators. No sense in doing this without doing it right. And I don't think ketmar will do it, so if @schuetzm isn't, I think we can just close it.

**Pull request closed without merging**